src/shared/selinux-util.c

```diff
@@ -434,40 +444,84 @@ int mac_selinux_get_child_mls_label(int socket_fd, const char *exe, const char *
+                if (!targetcon)
+                        return -ENOMEM;
+        } else {
                 /* If there is no context set for next exec let's use context
                    of target executable */
-                r = getfilecon_raw(exe, &fcon);
+                r = getfilecon_raw(exe, &targetcon);
                 if (r < 0)
                         return -errno;
         }
 
-        bcon = context_new(mycon);
-        if (!bcon)
+        /* If neither SELinuxContext= was set,
+         * nor a SELinux label exists on the executable file,
+         * use the context under which systemd is running */
+        if (!targetcon) {
+              targetcon = strdup(mycon);
+              if (!targetcon)
+                    return -ENOMEM;
+        }
+
+        tcon = context_new(targetcon);
+        if (!tcon)
                 return -ENOMEM;
 
+        /* Get context of the network socket */
         pcon = context_new(peercon);
         if (!pcon)
                 return -ENOMEM;
 
+        /* Extract last part of the network socket context
+         * (e.g.: systemd_u:object_r:netlabel_peer_t:s0 -> s0) */
         range = context_range_get(pcon);
         if (!range)
                 return -errno;
 
-        r = context_range_set(bcon, range);
+        /* Now change last part of previously computed targetcon to the
+         * one extracted from the network socket */
+        r = context_range_set(tcon, range);
         if (r)
                 return -errno;
 
-        freecon(mycon);
-        mycon = strdup(context_str(bcon));
-        if (!mycon)
-                return -ENOMEM;
-
         sclass = string_to_security_class("process");
         if (sclass == 0)
                 return -ENOSYS;
 
-        return RET_NERRNO(security_compute_create_raw(mycon, fcon, sclass, label));
+        resultcon = strdup(context_str(tcon));
+        if (!resultcon)
+                return -ENOMEM;
+
+        /* If SELinuxContext= was set explicitly, use it, replace MLS level
+         * with the one from socket and fail later if SELinux does not allow
+         * to transition to this context */
+        if (exec_label) {
+                *label = strdup(resultcon);
+                if (!*label)
+                      return -ENOMEM;
+        /* If heurestics were used, ask the Linux kernel which context it "recommends"
+         * (going here with explicitly set SELinuxContext= may result
+         * in its value silently being not applied) */
+        } else {
+                /* bcon is our current context, but with MLS level from socket;
+                 * required to calculate transition */
+                bcon = context_new(mycon);
+                if (!bcon)
+                        return -ENOMEM;
+                r = context_range_set(bcon, range);
+                if (r)
+                        return -errno;
+                freecon(mycon);
+                mycon = strdup(context_str(bcon));
```

Comment:
why make a copy of the string, if you just free it later?

Reply:
bcon is not a string (because context_range_set() works with context_t, not strings), it is context_t, but mycon is a string. A string is needed for security_compute_create_raw(), that is why I convert context_t bcon to char* mycon

Reply:
You are not just converting the context, you are then duplicating the result via strdup(), and there's really no reason to do so

Reply:
The only way how to avoid strdup() here that I see is the following:

```diff
diff --git a/src/basic/selinux-util.c b/src/basic/selinux-util.c
index 4ee3471640..23fd29ea20 100644
--- a/src/basic/selinux-util.c
+++ b/src/basic/selinux-util.c
@@ -348,11 +348,7 @@ int mac_selinux_get_child_mls_label(int socket_fd, const char *exe, const char *
                 r = context_range_set(bcon, range);
                 if (r)
                         return -errno;
-                freecon(mycon);
-                mycon = strdup(context_str(bcon));
-                if (!mycon)
-                        return -ENOMEM;
-                r = security_compute_create_raw(mycon, resultcon, sclass, label);
+                r = security_compute_create_raw(context_str(bcon), resultcon, sclass, label);
                 if (r < 0)
                         return -errno;
         }
```
It will remove control if context_str() succeeds
Am I missing something?

Reply:
yeah, sounds ok to me.

Reply:
or assing the result to a temporary vairable `const char *bcon_str` or so, if context_str() can fail.